lnrpc/lightning.proto

```diff
@@ -3898,3 +3898,187 @@ message Op {
     string entity = 1;
     repeated string actions = 2;
 }
+
+message RPCMiddlewareRequest {
+    /*
+    The unique ID of the intercepted request. Useful for mapping request to
+    response when implementing full duplex message interception.
+    */
+    uint64 request_id = 1;
+
+    /*
+    The raw bytes of the complete macaroon as sent by the gRPC client in the
+    original request.
+    */
+    bytes raw_macaroon = 2;
+
+    /*
+    The parsed condition of the macaroon's custom caveat for convenient access.
+    This field only contains the value of the custom caveat that the handling
+    middleware has registered itself for. The condition _must_ be validated for
+    messages of intercept_type stream_auth and request!
+    */
+    string custom_caveat_condition = 3;
+
+    /*
+    There are three types of messages that will be sent to the middleware for
+    inspection and approval: Stream authentication, request and response
+    interception. The first two can only be accepted (=forward to main RPC
+    server) or denied (=return error to client). Intercepted responses can also
+    be replaced/overwritten.
+    */
+    oneof intercept_type {
+        /*
+        Intercept stream authentication: each new streaming RPC call that is
+        initiated against lnd and contains the middleware's custom macaroon
+        caveat can be approved or denied based upon the macaroon in the stream
+        header. This message will only be sent for streaming RPCs, unary RPCs
+        must handle the macaroon authentication in the request interception to
+        avoid an additional message round trip between lnd and the middleware.
+        */
+        StreamAuth stream_auth = 4;
+
+        /*
+        Intercept incoming gRPC client request message: all incoming messages,
+        both on streaming and unary RPCs, are forwarded to the middleware for
+        inspection. For unary RPC messages the middleware is also expected to
```

Comment:
> "For unary RPC messages the middleware is also expected to validate the custom macaroon caveat of the request."

Wouldn't middleware also need to validate custom macaroon caveats for streaming calls?

Reply:
Yes. The difference is that for an unary call, the "setup" or "authorization" of an RPC call (when we "see" the call for the first time in the gRPC interceptor and want to decide whether to accept or refuse the call) and the arrival of the gRPC request message happens in the same call. That's why both the _authorization_ and the interception of the actual request message happen in the same step.
For a streaming call, the "setup" or "authorization" (decide whether to accept or refuse a stream initialization) can happen a long time before the first gRPC request message arrives (if ever). That's why it's split into two interception steps, so you can individually authorize (accept/refuse) a call, independent of whether a message is sent initially or not.
Hope that makes sense. Let me know if you think that needs to be reflected better in the comments.